src/storage/sql_table.cpp

```diff
@@ -1,20 +1,314 @@
+          const auto slot UNUSED_ATTRIBUTE = desired_v.data_table_->Insert(txn, *pr);
+        }
+      }
+      delete[] buffer;
+    }
+  }
+  if (!result) {
+    // For MVCC correctness, this txn must now abort for the GC to clean up the version chain in the DataTable
+    // correctly.
+    txn->SetMustAbort();
+  }
+  return std::make_pair(result, returned_slot);
+}
+
+TupleSlot SqlTable::Insert(const common::ManagedPointer<transaction::TransactionContext> txn, RedoRecord *const redo,
+                           layout_version_t layout_version) const {
+  TERRIER_ASSERT(redo->GetTupleSlot() == TupleSlot(nullptr, 0), "TupleSlot was set in this RedoRecord.");
+  TERRIER_ASSERT(redo == reinterpret_cast<LogRecord *>(txn->redo_buffer_.LastRecord())
+                             ->LogRecord::GetUnderlyingRecordBodyAs<RedoRecord>(),
+                 "This RedoRecord is not the most recent entry in the txn's RedoBuffer. Was StageWrite called "
+                 "immediately before?");
+  const auto slot = tables_.at(layout_version).data_table_->Insert(txn, *(redo->Delta()));
+  redo->SetTupleSlot(slot);
+  return slot;
+}
+
+void SqlTable::Scan(const terrier::common::ManagedPointer<transaction::TransactionContext> txn,
+                    DataTable::SlotIterator *const start_pos, ProjectedColumns *const out_buffer,
+                    const layout_version_t layout_version) const {
+  layout_version_t tuple_version = (*start_pos)->GetBlock()->layout_version_;
+
+  TERRIER_ASSERT(out_buffer->NumColumns() <= tables_.at(layout_version).column_oid_to_id_map_.size(),
+                 "The output buffer never returns the version pointer columns, so it should have "
+                 "fewer attributes.");
+
+  // Check for version match
+  if (tuple_version == layout_version) {
+    tables_.at(layout_version).data_table_->Scan(txn, start_pos, out_buffer);
+    return;
+  }
+
+  col_id_t ori_header[out_buffer->NumColumns()];
+  bool missing = false;
+  uint32_t filled = 0;
+  auto desired_v = tables_.at(layout_version);
+
+  for (layout_version_t i = tuple_version; i <= layout_version && out_buffer->NumTuples() < out_buffer->MaxTuples();
+       i++) {
+    auto tuple_v = tables_.at(i);
+    if (AlignHeaderToVersion(out_buffer, tuple_v, desired_v, &ori_header[0])) missing = true;
+
+    if (i != tuple_version) *start_pos = tuple_v.data_table_->begin();
+    tuple_v.data_table_->IncrementalScan(txn, start_pos, out_buffer, filled);
+    filled = out_buffer->NumTuples();
+  }
+
+  // copy back the original header
+  std::memcpy(out_buffer->ColumnIds(), ori_header, sizeof(col_id_t) * out_buffer->NumColumns());
+
+  if (missing) {
+    for (uint32_t idx = 0; idx < out_buffer->NumTuples(); idx++) {
+      ProjectedColumns::RowView row = out_buffer->InterpretAsRow(idx);
+      FillMissingColumns(&row, desired_v);
+    }
+  }
+}
+
+// return false if it is null
+static bool PeekValue(const type::TransientValue &transient_val, byte *value_output) {
```

Comment:
This should probably live in type::TransientValuePeeker, so it can be used elsewhere in the codebase